### ISSUE TITLE
Defining fields after additional functions causes mayhem in record

### ISSUE THREAD

**Issue by m0davis** [agda/agda#1720]
It's sometimes convenient to declare some fields in a record, define some additional functions, and only then define some more fields (that use those additional functions). Unfortunately, doing this right now is not without its headaches. Here are two small samples that demonstrate the problems I've found. In both of the below problems, moving the field declaration to the top avoids the errors. 

PROBLEM 1: "where" syntax not accepted

```
record Foo (A : Set) : Set where
  a : A
  a = {!!} where
    a' : A
    a' = {!!}

  field
    f : A
```

The error is:

```
Not a valid let-declaration
when scope checking
let mutual
      a : A
      a = ?
        where
          a' : A
          a' = ?
in (f : A) → Set₀
```

PROBLEM 2: Goal is not properly recognized

```
record Bar (A B : Set) : Set where
  a : A
  a = {!!} 

  b : B
  b = {!!}

  field
    f : A
```

Agda should recognize two goals but instead lists four. C-c C-, in function b shows that the goal is A, when it should be B.

**Comment by m0davis** [comment]
PROBLEM 3: postulate block not recognized

```
record Baz (A B : Set) : Set where
  postulate
    g : B

  field
    f : A
```

The error is:

```
Missing definition for g
```

**Comment by andreasabel** [comment]
Definitions in records that precede field declarations are not first-class, only simple definitions are allowed (e.g. `where` is forbidden) (PROBLEMS 1 and 2).  To improve the implementation is a long standing issue, see #1570, #917, #528, #434.  PROBLEM 3 is related, but the current behavior is clearly a bug.

**Comment by andreasabel** [comment]
PROBLEM 1 and PROBLEM 3 is fixed now, see work on #779.  You get a proper error now.

PROBLEM 2 remains open.
